tools/batch-script/address_provider.go

```diff
@@ -0,0 +1,224 @@
+		Uint("numAccounts", lastAddressIndex).
+		Int("stepsNeeded", searchStep).
+		Msg("Found last address")
+
+	ap.lastAddress = ap.indexToAddress(lastAddressIndex)
+	ap.lastAddressIndex = lastAddressIndex
+	return ap, nil
+}
+
+// getLastAddress is a recursive function that finds the last address. Will use max 2 * log2(number_of_addresses) steps
+// If the last address is at index 7 the algorithm goes like this:
+// (x,y) <- lower and upper index
+// 0. start at (1,2); address exists at 2
+// 1. (2,4): address exists at 4
+// 2. (4,8): address doesnt exist at 8
+// 3. (4,8): check address (8 - 4) / 2 = 6  address exists so next pair is (6,8)
+// 4. (6,8): check address 7 address exists so next pair is (7,8)
+// 5. (7,8): check address (8 - 7) / 2 = 7 ... ok already checked so this is the last existing address
+//
+func (p *AddressProvider) getLastAddress(
+	lowerIndex uint,
+	upperIndex uint,
+	upperExists bool,
+	addressExistsAtIndex func(uint) (bool, error),
+) (uint, error) {
+
+	// Does the address exist at upper bound?
+	if upperExists {
+		// double the upper bound, the current upper bound is now the lower upper bound
+		newUpperIndex := upperIndex * 2
+		newUpperExists, err := addressExistsAtIndex(newUpperIndex)
+		if err != nil {
+			return 0, err
+		}
+		return p.getLastAddress(upperIndex, newUpperIndex, newUpperExists, addressExistsAtIndex)
+	}
+
+	midIndex := (upperIndex-lowerIndex)/2 + lowerIndex
+	if midIndex == lowerIndex {
+		// we found the last address
+		return midIndex, nil
+	}
+
+	// Check if the address exists in the middle of the interval.
+	// If yes, then take the (mid, upper) as the next pair,
+	// else take (lower, mid) as the next pair
+	midIndexExists, err := addressExistsAtIndex(midIndex)
+	if err != nil {
+		return 0, err
+	}
+	if midIndexExists {
+		return p.getLastAddress(midIndex, upperIndex, upperExists, addressExistsAtIndex)
+	} else {
+		return p.getLastAddress(lowerIndex, midIndex, midIndexExists, addressExistsAtIndex)
+	}
+}
+
+func (p *AddressProvider) indexToAddress(index uint) flow.Address {
+	p.generator.SetIndex(index)
+	return p.generator.Address()
+}
+
+func (p *AddressProvider) GetNextAddress() (address flow.Address, isOutOfBounds bool) {
+	address = p.indexToAddress(p.currentIndex)
+
+	// Give some progress information every so often
+	if p.currentIndex%(p.lastAddressIndex/10) == 0 {
+		p.log.Info().Msgf("Processed %v %% accounts", p.currentIndex/(p.lastAddressIndex/10)*10)
+	}
+
+	if p.currentIndex > p.lastAddressIndex {
+		isOutOfBounds = true
+	}
+	p.currentIndex += 1
+
+	return
+}
+
+// These addresses are known to be broken on Mainnet
+var brokenAddresses = map[flow.Address]struct{}{
+	flow.HexToAddress("bf48a20670f179b8"): {},
+	flow.HexToAddress("5eba0297874a2bfd"): {},
+	flow.HexToAddress("474ec037bcd8accf"): {},
+	flow.HexToAddress("b0e80595d267f4eb"): {},
+}
+
+func (p *AddressProvider) GenerateAddressBatches(addressChan chan<- []flow.Address, batchSize int) {
+	var done bool
+	for {
```

Comment:
could use `!done` as the condition (while-loop version)
```suggestion
	for !done {
```